test/ln_test.py

```diff
@@ -6,121 +6,80 @@
-        self.log.info(self.warnet(f"ln rpc 0 payinvoice -f {inv}"))
-
-        self.wait_for_predicate(self.check_invoice_settled)
-
-        self.log.info("Ensuring channel-level channel policy settings: source")
-        payment = json.loads(self.warnet("ln rpc 0 listpayments"))["payments"][0]
-        assert (
-            payment["fee_msat"] == "5506"
-        ), f"Expected fee_msat to be 5506, got {payment['fee_msat']}"
-
-        self.log.info("Ensuring circuit breaker tracked payment")
-        assert len(self.get_cb_forwards(1)["forwards"]) == 1, "Expected one circuit breaker forward"
-
-    def test_ln_payment_2_to_0(self):
-        self.log.info("Test LN payment from 2 -> 0")
-        inv = json.loads(self.warnet("ln rpc 0 addinvoice --amt=1000"))["payment_request"]
-        self.log.info(f"Got invoice from node 0: {inv}")
-        self.log.info("Paying invoice from node 2...")
-        self.log.info(self.warnet(f"ln rpc 2 payinvoice -f {inv}"))
-
-        self.wait_for_predicate(lambda: self.check_invoices(0) == 1)
-
-        self.log.info("Ensuring channel-level channel policy settings: target")
-        payment = json.loads(self.warnet("ln rpc 2 listpayments"))["payments"][0]
-        assert (
-            payment["fee_msat"] == "2213"
-        ), f"Expected fee_msat to be 2213, got {payment['fee_msat']}"
-
-    def test_simln(self):
-        self.log.info("Engaging simln")
-        node2pub, _ = json.loads(self.warnet("ln rpc 2 getinfo"))["uris"][0].split("@")
-        activity = [
-            {"source": "ln-0", "destination": node2pub, "interval_secs": 1, "amount_msat": 2000}
-        ]
-        self.warnet(
-            f"network export --exclude=[1] --activity={json.dumps(activity).replace(' ', '')}"
-        )
-        self.wait_for_predicate(lambda: self.check_invoices(2) > 1)
-        assert self.check_invoices(0) == 1, "Expected one invoice for node 0"
-        assert self.check_invoices(1) == 0, "Expected no invoices for node 1"
-
-    def check_invoice_settled(self):
-        invs = json.loads(self.warnet("ln rpc 2 listinvoices"))["invoices"]
-        if len(invs) > 0 and invs[0]["state"] == "SETTLED":
-            self.log.info("Invoice settled")
-            return True
-        return False
-
-    def check_invoices(self, index):
-        invs = json.loads(self.warnet(f"ln rpc {index} listinvoices"))["invoices"]
-        settled = sum(1 for inv in invs if inv["state"] == "SETTLED")
-        self.log.debug(f"Node {index} has {settled} settled invoices")
-        return settled
+        graphs = []
+        for n in range(10):
+            ln = f"tank-{n:04d}-ln"
+            res = self.warnet(f"ln rpc {ln} describegraph")
+            graphs.append(json.loads(res)["edges"])
+
+        def check_policy(node: int, index: int, field: str, values: tuple):
+            self.log.info(f"Checking policy: Node={node} ch={index} Expected={field}:{values}")
+            graph = graphs[node]
+            assert len(graph) == 13
```

Comment:
This assertion appears to fail spuriously: https://github.com/mplsgrant/warnet/actions/runs/12283986192/job/34278744847#step:9:192

I image if we wait for the edges to materialize it will avoid this assertion error:

````python
        def ln_node_has_n_edges(ln: str, n: int, ) -> Optional[str]:
            res = self.warnet(f"ln rpc {ln} describegraph")
            edges = json.loads(res)["edges"]
            if len(edges) != n:
                return None
            return edges

        for n in range(10):
            ln = f"tank-{n:04d}-ln"
            partial_fn = partial(ln_node_has_n_edges, ln, 13)
            edges = self.wait_for_predicate(partial_fn)
            graphs.append(edges)

        def check_policy(node: int, index: int, field: str, values: tuple):
            self.log.info(f"Checking policy: Node={node} ch={index} Expected={field}:{values}")
            graph = graphs[node]
            assert len(graph) == 13
            ch = graph[index]
            a = int(ch["node1_policy"][field])
            b = int(ch["node2_policy"][field])
            assert values == (a, b) or values == (
                b,
                a,
            ), f"policy check failed:\nActual:\n{ch}\nExpected:\n{field}:{values}"
````

Side note: I also think we should let `wait_for_predicate` to return a value:

````python
    def wait_for_predicate(self, predicate, timeout=5 * 60, interval=5):
        self.log.debug(f"Waiting for predicate with timeout {timeout}s and interval {interval}s")
        while timeout > 0:
            try:
                result = predicate()
                if result:
                    return result
            except Exception:
                pass
            sleep(interval)
            timeout -= interval
        import inspect

        raise Exception(
            f"Timed out waiting for Truth from predicate: {inspect.getsource(predicate).strip()}"
````

Reply:
ln_init handles all the waiting, at this point in the test we should be able to guarantee the graph for every node. Hard to tell but in your test run it looks like ln_init didn't output anything... I'll watch for this in future test runs